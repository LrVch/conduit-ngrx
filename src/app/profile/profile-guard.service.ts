import { Injectable } from '@angular/core';
import {
    CanActivate,
    ActivatedRouteSnapshot
} from '@angular/router';
import { Observable, of, throwError } from 'rxjs';
import { Store } from '@ngrx/store';
import { AppState } from '../reducers';
import { Router } from '@angular/router';
import { tap, take, map, switchMap, catchError } from 'rxjs/operators';
import { ProfilesService } from '../core';
import { ProfileLoadSuccess } from './profile.actions';

@Injectable()
export class ProfileGuard implements CanActivate {
    NO_USER_NAME = 'No user name';
    constructor(
        private store: Store<AppState>,
        private router: Router,
        private profileService: ProfilesService
    ) {

    }
    canActivate(
        route: ActivatedRouteSnapshot
    ): Observable<boolean> {

        const username = route.params['username'];

        return this.store.pipe(
            take(1),
            switchMap(() => username ? this.profileService.get(username) : throwError(this.NO_USER_NAME)),
            map(profile => new ProfileLoadSuccess({ profile })),
            tap((action: ProfileLoadSuccess) => this.store.dispatch(action)),
            map(profile => !!profile),
            catchError((error) => {
                console.error(error);
                this.router.navigate(['/']);
                return of(false);
            })
        );
    }
}
