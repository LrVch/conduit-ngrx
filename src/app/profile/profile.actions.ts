import { Action } from '@ngrx/store';
import { Profile } from '@app/core';

export enum ProfileActionTypes {
  ProfileLoadRequest = '[Profile] Profile Load Request',
  ProfileLoadSuccess = '[Profile] Profile Load Success',
  ProfileLoadFail = '[Profile] Profile Load Fail',
  ProfileClear = '[Profile] Profile Clear',

  ProfileToggleFollowingRequest = '[Profile] Profile Toggle Following Request',
  ProfileToggleFollowingSuccess = '[Profile] Profile Toggle Following Success',
  ProfileToggleFollowingFail = '[Profile] Profile Toggle Following Fail',
  ProfileFollowingRequest = '[Profile] Profile Following Request',
  ProfileUnFollowingRequest = '[Profile] Profile UnFollowing Request',

  SetFollowingProfile = '[Profile] Set Following Profile',
  ClearFollowingProfile = '[Profile] Clear Following Profile'
}

export class ProfileLoadRequest implements Action {
  readonly type = ProfileActionTypes.ProfileLoadRequest;

  constructor(public payload: { username: string }) { }
}

export class ProfileLoadSuccess implements Action {
  readonly type = ProfileActionTypes.ProfileLoadSuccess;

  constructor(public payload: { profile: Profile }) { }
}

export class ProfileLoadFail implements Action {
  readonly type = ProfileActionTypes.ProfileLoadFail;
}

export class ProfileClear implements Action {
  readonly type = ProfileActionTypes.ProfileClear;
}

export class ProfileToggleFollowingRequest implements Action {
  readonly type = ProfileActionTypes.ProfileToggleFollowingRequest;

  constructor(public payload: { profile: Profile }) { }
}

export class ProfileToggleFollowingSuccess implements Action {
  readonly type = ProfileActionTypes.ProfileToggleFollowingSuccess;

  constructor(public payload: { profile: Profile, showNotification?: boolean }) { }
}

export class ProfileToggleFollowingFail implements Action {
  readonly type = ProfileActionTypes.ProfileToggleFollowingFail;

  constructor(public payload?: { profile: Profile, showNotification?: boolean }) { }
}

export class ProfileFollowingRequest implements Action {
  readonly type = ProfileActionTypes.ProfileFollowingRequest;

  constructor(public payload: { profile: Profile }) { }
}

export class ProfileUnFollowingRequest implements Action {
  readonly type = ProfileActionTypes.ProfileUnFollowingRequest;

  constructor(public payload: { profile: Profile }) { }
}

export class SetFollowingProfile implements Action {
  readonly type = ProfileActionTypes.SetFollowingProfile;

  constructor(public payload: { profile: Profile }) { }
}

export class ClearFollowingProfile implements Action {
  readonly type = ProfileActionTypes.ClearFollowingProfile;
}

export type ProfileActions =
  ProfileLoadRequest
  | ProfileLoadSuccess
  | ProfileLoadFail
  | ProfileClear
  | ProfileToggleFollowingRequest
  | ProfileToggleFollowingSuccess
  | ProfileToggleFollowingFail
  | ProfileFollowingRequest
  | ProfileUnFollowingRequest
  | SetFollowingProfile
  | ClearFollowingProfile;
