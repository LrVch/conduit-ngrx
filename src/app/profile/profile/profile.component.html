<div class="profile-page">

  <app-user-info (folowing)="onFolowing($event)"
                 [isUser]="isUser$ | async"
                 [profile]="profile$ | async"></app-user-info>

  <div class="content">

    <div class="container">
      <div class="row">
        <div class="col-xs-12 col-md-10 offset-md-1">
          <app-content-tabs [type]="type$ | async"
                            [tabs]="tabs$ | async"
                            (selectedType)="onSelectedType($event)">
          </app-content-tabs>
        </div>
      </div>
    </div>

    <div class="container">
      <div class="row">
        <div class="col-xs-12 col-md-10 offset-md-1">
          <div class="articles">
            <app-articles-list [isLoading]="loadingArticles$ | async"
                               [locale]="locale$ | async"
                               [contentLength]="limit$ | async"
                               [isErrorLoading]="errorLoadingArticles$ | async"
                               (favorited)="onFavorited($event)"
                               [articlesList]="atricles$ | async"></app-articles-list>
          </div>
        </div>
      </div>

    </div>
    <div class="container">
      <div class="row">
        <div class="col-xs-12 col-md-10 offset-md-1">
          <mat-paginator [hidden]="(articlesCount$ | async) < 1 || (loadingArticles$ | async)"
                         [length]="articlesCount$ | async"
                         (page)="onPageChange($event)"
                         [pageSize]="limit$ | async"
                         [pageIndex]="pageIndex$ | async"
                         [showFirstLastButtons]="true"
                         [pageSizeOptions]="pageSizeOptions">
          </mat-paginator>
        </div>
      </div>
      <div class="pagination">
      </div>
    </div>
  </div>