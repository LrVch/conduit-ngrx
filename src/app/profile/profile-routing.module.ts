import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { AuthGuard } from '@app/core';
import { ProfileComponent } from './profile/profile.component';
import { ProfileGuard } from './profile-guard.service';

const routes: Routes = [
  {
    path: '',
    component: ProfileComponent,
    canActivate: [ProfileGuard],
    data: { title: 'conduit.menu.profile' }
  },
  {
    path: ':username',
    component: ProfileComponent,
    canActivate: [ProfileGuard],
    data: { title: 'conduit.menu.profile' }
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class ProfileRoutingModule {}
