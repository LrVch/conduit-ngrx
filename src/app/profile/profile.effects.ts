import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { Router } from '@angular/router';
import {
  ProfileActionTypes,
  ProfileToggleFollowingRequest,
  ProfileUnFollowingRequest,
  ProfileFollowingRequest,
  ProfileToggleFollowingSuccess,
  ProfileToggleFollowingFail,
  SetFollowingProfile,
  ClearFollowingProfile
} from './profile.actions';
import { retry, catchError, map, filter, mergeMap, withLatestFrom, tap } from 'rxjs/operators';
import { ProfilesService, Profile } from '../core';
import { of, Observable } from 'rxjs';
import { selectAuthLoggedIn } from '../auth/auth.selectors';
import { LogoutAction, ClearReturnStateFromRouteChange, AuthActionTypes, LoginSuccess } from '../auth/auth.actions';
import { selectFollowingProfile } from './profile.selectors';
import { NotificationService } from '../core/services/notification.service';


@Injectable()
export class ProfileEffects {

  constructor(
    private actions$: Actions,
    private store: Store<AppState>,
    private profileService: ProfilesService,
    private notificationService: NotificationService
  ) { }

  @Effect()
  taggleFollowUserProfile$ = this.actions$.pipe(
    ofType<ProfileToggleFollowingRequest>(ProfileActionTypes.ProfileToggleFollowingRequest),
    withLatestFrom(this.store.select(selectAuthLoggedIn)),
    mergeMap(([action, isLoggedIn]): Observable<ProfileUnFollowingRequest | ProfileFollowingRequest | SetFollowingProfile> => {
      if (!isLoggedIn) {
        return of(new SetFollowingProfile({ profile: action.payload.profile }));
      }
      const { profile } = action.payload;
      const { following } = profile;

      if (following) {
        return of(new ProfileUnFollowingRequest({ profile }));
      } else {
        return of(new ProfileFollowingRequest({ profile }));
      }
    })
  );

  @Effect()
  setFollowingProfile$ = this.actions$.pipe(
    ofType<SetFollowingProfile>(ProfileActionTypes.SetFollowingProfile),
    map(() => new LogoutAction())
  );

  @Effect()
  clearReturnStateFromRouteChange$ = this.actions$.pipe(
    ofType<ClearReturnStateFromRouteChange>(AuthActionTypes.ClearReturnStateFromRouteChange),
    map(() => new ClearFollowingProfile())
  );

  @Effect()
  loginSuccessProfile$ = this.actions$.pipe(
    ofType<LoginSuccess>(AuthActionTypes.LoginSuccess),
    withLatestFrom(this.store.select(selectFollowingProfile)),
    filter(([action, profile]) => !!profile),
    map(([action, profile]) => new ProfileFollowingRequest({ profile }))
  );

  @Effect()
  profileFollowing$ = this.actions$.pipe(
    ofType<ProfileFollowingRequest>(ProfileActionTypes.ProfileFollowingRequest),
    withLatestFrom(this.store.select(selectFollowingProfile)),
    mergeMap(([action, favoritingProfile]) => {
      const { profile } = action.payload;
      const { username } = profile;

      return this.profileService.follow(username)
        .pipe(
          map(result => {
            return new ProfileToggleFollowingSuccess({ profile: result.profile, showNotification: !favoritingProfile });
          }),
          retry(3),
          catchError(error => {
            console.error(error);
            return of(new ProfileToggleFollowingFail({showNotification: !favoritingProfile}));
          })
        );
    })
  );

  @Effect()
  profileUnFollowing$ = this.actions$.pipe(
    ofType<ProfileUnFollowingRequest>(ProfileActionTypes.ProfileUnFollowingRequest),
    mergeMap((action) => {
      const { profile } = action.payload;
      const { username } = profile;

      return this.profileService.unfollow(username)
        .pipe(
          map(result => {
            return new ProfileToggleFollowingSuccess({ profile: result.profile });
          }),
          retry(3),
          catchError(error => {
            console.error(error);
            return of(new ProfileToggleFollowingFail());
          })
        );
    })
  );

  @Effect({ dispatch: false })
  profileToggleFollowingSuccess$ = this.actions$.pipe(
    ofType<ProfileToggleFollowingSuccess>(ProfileActionTypes.ProfileToggleFollowingSuccess),
    filter(action => action.payload.showNotification),
    map(action => action.payload.profile),
    tap((profile: Profile) => this.notificationService.success({ message: 'Added to your favorites authors' }))
  );

  @Effect({ dispatch: false })
  profileToggleFollowingFail$ = this.actions$.pipe(
    ofType<ProfileToggleFollowingFail>(ProfileActionTypes.ProfileToggleFollowingFail),
    filter(action => action.payload.showNotification),
    tap(() => this.notificationService.error({ message: 'Can\'t add author to your favorites' }))
  );
}
