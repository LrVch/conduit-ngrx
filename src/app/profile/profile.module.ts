import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ProfileComponent } from './profile/profile.component';
import { ProfileRoutingModule } from './profile-routing.module';
import { SharedModule } from '@app/shared';
import { EffectsModule } from '@ngrx/effects';
import { ProfileEffects } from './profile.effects';
import { StoreModule } from '@ngrx/store';
import * as fromProfile from './profile.reducer';
import { ProfileGuard } from './profile-guard.service';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { HttpClient } from '@angular/common/http';
import { environment } from '@env/environment';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    ProfileRoutingModule,
    EffectsModule.forFeature([ProfileEffects]),
    StoreModule.forFeature('profile', fromProfile.profileReducer),
    TranslateModule.forChild({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient]
      }
    })
  ],
  declarations: [ProfileComponent],
  providers: [ProfileGuard]
})
export class ProfileModule {}

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.i18nPrefix}/assets/i18n/`,
    '.json'
  );
}
