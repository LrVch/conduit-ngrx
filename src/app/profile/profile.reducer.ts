import { Profile } from '@app/core';
import { ProfileActions, ProfileActionTypes } from './profile.actions';

export interface ProfileState {
  profile: Profile;
  followingProfile: Profile | null;
}

export const initialState: ProfileState = {
  profile: null,
  followingProfile: null
};

export function profileReducer(
  state = initialState,
  action: ProfileActions
): ProfileState {
  switch (action.type) {
    case ProfileActionTypes.ProfileLoadSuccess:
      return {
        ...state,
        profile: action.payload.profile
      };
    case ProfileActionTypes.ProfileClear:
      return {
        ...state,
        profile: null
      };
    case ProfileActionTypes.ProfileToggleFollowingRequest:
      return {
        ...state,
        profile: { ...state.profile, isFollowLoading: true }
      };
    case ProfileActionTypes.ProfileToggleFollowingFail:
      return {
        ...state,
        profile: { ...state.profile, isFollowLoading: false }
      };
    case ProfileActionTypes.ProfileToggleFollowingSuccess:
      return {
        ...state,
        profile: action.payload.profile
      };
    case ProfileActionTypes.SetFollowingProfile:
      return {
        ...state,
        followingProfile: action.payload.profile
      };
    case ProfileActionTypes.ClearFollowingProfile:
      return {
        ...state,
        followingProfile: null
      };
    default:
      return state;
  }
}
