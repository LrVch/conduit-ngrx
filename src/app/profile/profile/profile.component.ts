import { Component, OnInit, OnDestroy } from '@angular/core';
import { Observable, of, Subject, combineLatest } from 'rxjs';
import { withLatestFrom, map, switchMap, take, filter, takeUntil } from 'rxjs/operators';
import { Profile, User, Article } from '@app/core';
import { AppState } from '@app/reducers';
import { Store, select } from '@ngrx/store';
import { selectUser } from '@app/auth/auth.selectors';
import { ProfileClear, ProfileToggleFollowingRequest } from '@app/profile/profile.actions';
import { selectProfile, selectProfileUsername } from '@app/profile/profile.selectors';
import {
  ResetConfig,
  SetAuthor,
  LoadArticlesRequest,
  ToggleArticleFavoriteRequest,
  SetLimit, SetOffset,
  SetFavorited,
  SetPageIndex
} from '@app/articles/articles.actions';
import {
  selectArticlesItems,
  getArticlesLimit,
  selectArticlesCount,
  selectArticlesLoading,
  getArticlesAuthor,
  getArticlesPageIndex,
  selectArticlesReturnConfig
} from '@app/articles/articles.selectors';
import { ArticlesConfigState } from '@app/articles/articlesConfig.reducer';

@Component({
  selector: 'app-profile',
  templateUrl: './profile.component.html'
})
export class ProfileComponent implements OnInit, OnDestroy {
  isUser$: Observable<boolean>;
  profile$: Observable<Profile>;
  tags$: Observable<Array<string>>;
  limit$: Observable<number>;
  pageIndex$: Observable<number>;
  atricles$: Observable<Article[]>;
  type$: Observable<string>;
  tabs = [{ title: 'By Author', value: 'author' }, { title: 'Favorited Posts', value: 'favorited' }];
  pageSizeOptions = [5, 10, 25, 100];
  articlesCount$: Observable<number>;
  loadingArticles$: Observable<boolean>;
  errorLoadingArticles$: Observable<boolean>;
  destroStream$ = new Subject<any>();

  constructor(
    private store: Store<AppState>
  ) { }

  ngOnInit() {
    this.profile$ = this.store.pipe(select(selectProfile));
    this.isUser$ = this.profile$.pipe(
      withLatestFrom(this.store.pipe(select(selectUser))),
      filter(([profile, user]: [Profile, User]) => !!profile),
      switchMap(([profile, user]: [Profile, User]) => {
        return user ? of(profile.username === user.username) : of(false);
      }),
    );

    this.atricles$ = this.store.pipe(select(selectArticlesItems));
    this.limit$ = this.store.pipe(select(getArticlesLimit));
    this.pageIndex$ = this.store.pipe(select(getArticlesPageIndex));
    this.articlesCount$ = this.store.pipe(select(selectArticlesCount));
    this.loadingArticles$ = this.store.pipe(select(selectArticlesLoading));

    combineLatest(
      this.store.pipe(select(selectArticlesReturnConfig)),
      this.store.pipe(select(selectProfileUsername))
    ).pipe(
      filter(([config, username]: [ArticlesConfigState, string]) => !!username),
      takeUntil(this.destroStream$)
    ).subscribe(([config, username]: [ArticlesConfigState, string]) => {
      this.store.dispatch(new ResetConfig());
      if (config) {
        const { author, favorited, limit, offset } = config.filters;
        this.store.dispatch(new SetAuthor({ author }));
        this.store.dispatch(new SetFavorited({ favorited }));
        this.store.dispatch(new SetLimit({ limit }));
        this.store.dispatch(new SetOffset({ offset }));
        this.store.dispatch(new SetPageIndex({ pageIndex: offset / limit }));
      } else {
        this.store.dispatch(new SetAuthor({ author: username }));
      }
      this.store.dispatch(new LoadArticlesRequest());
    });

    this.type$ = this.store.pipe(select(getArticlesAuthor),
      map(author => author ? 'author' : 'favorited'));
  }

  ngOnDestroy() {
    this.store.dispatch(new ProfileClear());
    this.destroStream$.next();
    this.destroStream$.complete();
  }

  onFolowing(profile: Profile): void {
    this.store.dispatch(new ProfileToggleFollowingRequest({ profile }));
  }

  onSelectedType(type: string): void {
    this.store.dispatch(new ResetConfig());
    this.store.pipe(
      select(selectProfile),
      map(profile => profile.username),
      take(1)
    ).subscribe(username => {
      if (type === 'author') {
        this.store.dispatch(new SetAuthor({ author: username }));
      } else {
        this.store.dispatch(new SetFavorited({ favorited: username }));
      }
      this.store.dispatch(new LoadArticlesRequest());
    });

  }

  onPageChange(page): void {
    const { pageIndex, pageSize } = page;
    const limit = pageSize;
    const offset = limit * pageIndex;
    this.store.dispatch(new SetLimit({ limit }));
    this.store.dispatch(new SetPageIndex({ pageIndex }));
    this.store.dispatch(new SetOffset({ offset }));
    this.store.dispatch(new LoadArticlesRequest());
  }

  onFavorited(article: Article): void {
    this.store.dispatch(new ToggleArticleFavoriteRequest({ article }));
  }
}
