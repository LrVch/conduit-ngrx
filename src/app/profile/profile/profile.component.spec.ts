import { async, ComponentFixture, fakeAsync, tick } from '@angular/core/testing';
import { ConfigureFn, configureTests, getUser, getProfile, getArticle } from '@app/lib/testing';
import { ProfileComponent } from './profile.component';
import { DebugElement, Component, Output, EventEmitter, Input } from '@angular/core';
import { Errors, User, Profile, Article } from '@app/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';
import { Store, StoreModule, combineReducers } from '@ngrx/store';
import * as ProfileActions from '../profile.actions';
import * as ArticlesActions from '@app/articles/articles.actions';
import * as AuthActions from '@app/auth/auth.actions';
import * as fromRoot from '@app/reducers';
import * as fromProfile from '@app/profile/profile.reducer';
import * as fromArticles from '@app/articles/articles.reducer';
import * as fromAuth from '@app/auth/auth.reducer';
import { hot, cold } from 'jasmine-marbles';
import { ArticlesConfigState } from '@app/articles/articlesConfig.reducer';
import { of } from 'rxjs';


@Component({
    selector: 'app-user-info',
    template: `<div id="follow" (click)="onFolowing()"></div>`
})
class TestHostUserInfoComponent {
    @Input('isUser') isUser: User;
    @Input('profile') profile: Profile;
    @Output() folowing = new EventEmitter<Profile>();

    onFolowing() {
        this.folowing.emit(getProfile());
    }
}

@Component({
    selector: 'app-content-tabs',
    template: `<div id="tabs" (click)="onSelectedType($event)"></div>`
})
class TestHostContentTabsComponent {
    @Input('type') type: any;
    @Input('tabs') tabs: any;
    @Output() selectedType = new EventEmitter();

    onSelectedType(event) {
        this.selectedType.emit(event.type);
    }
}

@Component({
    selector: 'app-articles-list',
    template: `<div id="favorite" (click)="onFavorited()"></div>`
})
class TestHostArticleListComponent {
    @Input('isLoading') isLoading: boolean;
    @Input('isErrorLoading') isErrorLoading: boolean;
    @Input('articlesList') articlesList: any;
    @Output() favorited = new EventEmitter();

    onFavorited() {
        this.favorited.emit(getArticle());
    }
}

@Component({
    selector: 'mat-paginator',
    template: `<div id="page" (click)="onPageChange($event)"></div>`
})
class TestHostMatPaginatorComponent {
    @Input('hidden') hidden: boolean;
    @Input('length') length: number;
    @Input('pageSize') pageSize: any;
    @Input('pageIndex') pageIndex: any;
    @Input('showFirstLastButtons') showFirstLastButtons: any;
    @Input('pageSizeOptions') pageSizeOptions: any;
    @Output() page = new EventEmitter();

    onPageChange(event) {
        this.page.emit(event.page);
    }
}


describe('ProfileComponent', () => {
    let component: ProfileComponent;
    let fixture: ComponentFixture<ProfileComponent>;
    let de: DebugElement;
    let store: Store<fromRoot.AppState>;
    let profile: Profile;
    let user: User;
    let article: Article;

    const returnConfig: ArticlesConfigState = {
        type: 'all',
        filters: {
            tag: 'hello',
            author: '',
            favorited: '',
            offset: 0,
            limit: 10,
            pageIndex: 0
        }
    };

    beforeEach(
        async(() => {
            profile = getProfile();
            user = getUser();
            article = getArticle();
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [
                        ProfileComponent,
                        TestHostUserInfoComponent,
                        TestHostArticleListComponent,
                        TestHostMatPaginatorComponent,
                        TestHostContentTabsComponent
                    ],
                    imports: [
                        NoopAnimationsModule,
                        StoreModule.forRoot({
                            ...fromRoot.reducers,
                            feature: combineReducers(
                                fromProfile.profileReducer,
                            ),
                        }),
                    ],
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(ProfileComponent);
                component = fixture.componentInstance;
                de = fixture.debugElement;
                store = testBed.get(Store);
                fixture.detectChanges();

                spyOn(store, 'dispatch').and.callThrough();
                spyOn(component.destroStream$, 'next').and.callThrough();
                spyOn(component.destroStream$, 'complete').and.callThrough();
            });
        })
    );

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should compile', () => {
        (<any>expect(fixture)).toMatchSnapshot();
    });


    it('should select profile', () => {
        component.profile$.subscribe(res => {
            expect(res).toBeNull();
        });
    });

    it('should select isUser and return true', () => {
        user.username = 'testuser';
        profile.username = 'testuser';

        store.dispatch(new AuthActions.LoginSuccess({ user }));
        store.dispatch(new ProfileActions.ProfileLoadSuccess({ profile }));

        component.ngOnInit();

        component.isUser$.pipe().subscribe(res => {
            expect(res).toBeTruthy();
        });
    });

    it('should select isUser and return false', () => {
        user.username = 'nottestuser';
        profile.username = 'testuser';

        store.dispatch(new AuthActions.LoginSuccess({ user }));
        store.dispatch(new ProfileActions.ProfileLoadSuccess({ profile }));

        component.ngOnInit();

        component.isUser$.pipe().subscribe(res => {
            expect(res).toBeFalsy();
        });
    });

    it('should select atricles', (done) => {
        component.atricles$.subscribe(res => {
            expect(res).toEqual([]);
            done();
        }, done, done);
    });

    it('should select limit', (done) => {
        component.limit$.subscribe(res => {
            expect(res).toBe(10);
            done();
        }, done, done);
    });

    it('should select pageIndex', (done) => {
        component.pageIndex$.subscribe(res => {
            expect(res).toBe(0);
            done();
        }, done, done);
    });

    it('should select articlesCount', (done) => {
        component.articlesCount$.subscribe(res => {
            expect(res).toBe(0);
            done();
        }, done, done);
    });

    it('should select loadingArticles', (done) => {
        component.loadingArticles$.subscribe(res => {
            expect(res).toBeFalsy();
            done();
        }, done, done);
    });

    it('should dispatch "ResetConfig, SetAuthor, LoadArticlesRequest" if there is no return articles config', () => {
        store.dispatch(new ProfileActions.ProfileLoadSuccess({ profile }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetAuthor({ author: 'username' }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
    });

    it('should dispatch set articles config if there is return articles config', () => {
        store.dispatch(new ProfileActions.ProfileLoadSuccess({ profile }));
        store.dispatch(new ArticlesActions.SetReturnArticlesConfig({ config: returnConfig }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
        const { author, favorited, limit, offset } = returnConfig.filters;
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetAuthor({ author }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetFavorited({ favorited }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetLimit({ limit }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetOffset({ offset }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetPageIndex({ pageIndex: offset / limit }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
    });

    it('should select type', (done) => {
        component.type$.subscribe(res => {
            expect(res).toBe('favorited');
            done();
        }, done, done);
    });

    it('should handle ngOnDestroy', () => {
        component.ngOnDestroy();
        expect(store.dispatch).toHaveBeenCalledWith(new ProfileActions.ProfileClear());
        expect(component.destroStream$.next).toHaveBeenCalled();
        expect(component.destroStream$.complete).toHaveBeenCalled();
    });


    it('should dispatch "ProfileToggleFollowingRequest"', () => {
        const followDe = de.query(By.css('#follow'));

        followDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new ProfileActions.ProfileToggleFollowingRequest({ profile }));
    });

    it('should dispatch "ToggleArticleFavoriteRequest"', () => {
        const favoriteDe = de.query(By.css('#favorite'));

        favoriteDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ToggleArticleFavoriteRequest({ article }));
    });

    it('should load article for sertain tab and set author', () => {
        const spy = jest.spyOn(store, 'pipe').mockImplementation(() => of('username'));
        const tabDe = de.query(By.css('#tabs'));
        tabDe.triggerEventHandler('click', { type: 'author' });

        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetAuthor({ author: 'username' }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
    });

    it('should load article for sertain tab and set favorited', () => {
        const spy = jest.spyOn(store, 'pipe').mockImplementation(() => of('username'));
        const tabDe = de.query(By.css('#tabs'));
        tabDe.triggerEventHandler('click', { type: 'favorited' });

        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetFavorited({ favorited: 'username' }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
    });

    it('should change page', () => {
        const page = {
            pageIndex: 20,
            pageSize: 20
        };
        const pageDe = de.query(By.css('#page'));
        pageDe.triggerEventHandler('click', { page });

        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetLimit({ limit: page.pageSize }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetPageIndex({ pageIndex: page.pageIndex }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetOffset({ offset: (page.pageSize * page.pageIndex) }));
        expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
    });
});
