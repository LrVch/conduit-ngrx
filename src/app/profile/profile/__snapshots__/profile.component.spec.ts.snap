// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileComponent should compile 1`] = `
<app-profile
  articlesCount$={[Function Store]}
  atricles$={[Function Store]}
  canModify$={[Function Observable]}
  destroStream$={[Function Subject]}
  isUser$={[Function Store]}
  limit$={[Function Store]}
  loadingArticles$={[Function Store]}
  locale$={[Function Store]}
  pageIndex$={[Function Store]}
  pageSizeOptions={[Function Array]}
  profile$={[Function Store]}
  store={[Function Store]}
  tabs$={[Function Observable]}
  translateService={[Function MockTranslateService]}
  type$={[Function Store]}
>
  <div
    class="profile-page"
  >
    <app-user-info>
      <div
        id="follow"
      />
    </app-user-info>
    <div
      class="content"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-xs-12 col-md-10 offset-md-1"
          >
            <app-content-tabs
              ng-reflect-tabs="[object Object],[object Object"
              ng-reflect-type="favorited"
            >
              <div
                id="tabs"
              />
            </app-content-tabs>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-xs-12 col-md-10 offset-md-1"
          >
            <div
              class="articles"
            >
              <app-articles-list
                ng-reflect-articles-list=""
                ng-reflect-content-length="10"
                ng-reflect-is-loading="false"
                ng-reflect-locale=""
              >
                <div
                  id="favorite"
                />
              </app-articles-list>
            </div>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-xs-12 col-md-10 offset-md-1"
          >
            <mat-paginator
              ng-reflect-hidden="true"
              ng-reflect-length="0"
              ng-reflect-page-index="0"
              ng-reflect-page-size="10"
              ng-reflect-page-size-options="5,10,25,100"
              ng-reflect-show-first-last-buttons="true"
            >
              <div
                id="page"
              />
            </mat-paginator>
          </div>
        </div>
        <div
          class="pagination"
        />
      </div>
    </div>
  </div>
</app-profile>
`;
